### PR TITLE
Add DistArray initializer_list ctor that supports tiling

### PR DESCRIPTION
**Pull request by ryanmrichard**
This is basically just the generalization of #174 to multiple tiles.

To properly unit test some functions I really need there to be tiles in the tensors. This PR will add additional ctors of the form:

```.cpp
DistArray(World&, const trange_type&, T);
```
(where `T` is one of the `initializer_list` types). The provided `initializer_list` is expected to be the entire tensor, properly nested in row-major form. The ctor will then pluck the elements out of the provided `initializer_list` and put them into the correct tiles.

TODO:

- [x] unit test and document `get_elem_from_il`
- [x] unit test and document new `array_from_il` overload
- [x] add, document, and unit test new `DistArray` ctors

### CODE REVIEW

**Comment by ryanmrichard** [comment]
Would love some C++ guru help. I thought I was clever and figured out how to condense the large number of typedefs for nested initializer lists by defining:

```.cpp
template<std::size_t N, typename T>
struct TensorILHelper {
  using type = typename TensorILHelper<N - 1, std::initializer_list<T>>::type;
};

template<typename T>
struct TensorILHelper<0, T> {
  using type = T;
};

template<std::size_t N, typename T>
using tensor_il = typename TensorILHelper<N, T>::type;
```

you get a helper typedef `tensor_il` which takes two template parameters: `N` the level of nesting and `T` the type of the elements in the most nested initializer list. You can verify that this works with:

```.cpp
static_assert(std::is_same_v<tensor_il<0, double>, double>);
static_assert(std::is_same_v<tensor_il<1, double>, std::initializer_list<double>>);
static_assert(std::is_same_v<tensor_il<2, double>, std::initializer_list<std::initializer_list<double>>>);
```

If i try to rewrite the `DistArray` ctors like:
```
template<std::size_t N, typename T>
DistArray(World& world, tensor_il<N, T> il);
```

and call them like:

```.cpp
tensor_il<2, double> il{...};
DistArray(world, il);
```
I get an error can't deduce N. I then tried rewriting the 6 ctors only parameterized on the element type:

```
template<typename T>
DistArray(World& world, tensor_il<1, T> il);
```
and it complains it can't deduce T.

I fail to see how the last one is any different than what is currently in master.

**Comment by evaleev** [comment]
> 
> I fail to see how the last one is any different than what is currently in master.

I don't think template parameter can be deduced from recursively defined types. The compiler has to stop somewhere. In general it would have to potentially expand the template definition arbitrarily deeply.
https://en.cppreference.com/w/cpp/language/template_argument_deduction does mention recursive, but I don't know if in practice compilers do the right thing.

TL;DR version: templates != macros .

**Comment by ryanmrichard** [comment]
Assuming the most recent push passes CI and @evaleev sees nothing wrong with it, this is r2g.
